nodemanagement/node_management_worker.go

```diff
@@ -0,0 +1,326 @@
+			} else if existingStatus == nil {
+				newStatus := exchangecommon.StatusFromNewPolicy(policy)
+				org, nodeId := cutil.SplitOrgSpecUrl(n.GetExchangeId())
+				if err = persistence.SaveOrUpdateNMPStatus(n.db, name, newStatus); err != nil {
+					return err
+				} else if _, err = exchange.PutNodeManagementPolicyStatus(n, org, nodeId, name, &newStatus); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (n *NodeManagementWorker) NewEvent(incoming events.Message) {
+	glog.Infof(nmwlog(fmt.Sprintf("Handling event: %v", incoming)))
+	switch incoming.(type) {
+	case *events.EdgeRegisteredExchangeMessage:
+		msg, _ := incoming.(*events.EdgeRegisteredExchangeMessage)
+
+		switch msg.Event().Id {
+		case events.NEW_DEVICE_REG:
+			cmd := NewNodeRegisteredCommand(msg)
+			n.Commands <- cmd
+		}
+	case *events.NodeShutdownCompleteMessage:
+		msg, _ := incoming.(*events.NodeShutdownCompleteMessage)
+		switch msg.Event().Id {
+		case events.UNCONFIGURE_COMPLETE:
+			n.Commands <- worker.NewTerminateCommand("shutdown")
+		}
+	case *events.NodeShutdownMessage:
+		msg, _ := incoming.(*events.NodeShutdownMessage)
+		cmd := NewNodeShutdownCommand(msg)
+		n.Commands <- cmd
+	case *events.NMPDownloadCompleteMessage:
+		msg, _ := incoming.(*events.NMPDownloadCompleteMessage)
+
+		switch msg.Event().Id {
+		case events.NMP_DOWNLOAD_COMPLETE:
+			cmd := NewNMPDownloadCompleteCommand(msg)
+			n.Commands <- cmd
+		}
+	}
+}
+
+func VerifyCompatible(nodePol *externalpolicy.ExternalPolicy, nodePattern string, nmPol *exchangecommon.ExchangeNodeManagementPolicy) (bool, error) {
+	if nodePattern != "" {
+		return cutil.SliceContains(nmPol.Patterns, nodePattern), nil
+	}
+	if err := nodePol.Constraints.IsSatisfiedBy(nmPol.Properties); err != nil {
+		return false, err
+	} else if err = nmPol.Constraints.IsSatisfiedBy(nodePol.Properties); err != nil {
+		return false, err
+	} else {
+		return true, nil
+	}
+}
+
+func (n *NodeManagementWorker) CheckNMPStatus(baseWorkingFile string, statusFileName string) error {
+	/*
+		Check working dir for folders
+		for each folder
+			collect status
+			update exchange status
+			if sucessfull
+				remove folder
+	*/
+	files, err := ioutil.ReadDir(baseWorkingFile)
+	if err != nil {
+		return fmt.Errorf("Error reading the management worker files: %v", err)
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			if fileInfo, err := os.Stat(path.Join(baseWorkingFile, file.Name(), statusFileName)); err == nil && fileInfo != nil {
+				if err = n.CollectStatus(path.Join(baseWorkingFile, file.Name()), statusFileName, file.Name()); err != nil {
+					return err
+				}
+			} else if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+type StatusFile struct {
+	Status         string `json:"status"`
+	StartTime      uint64 `json:"startTime"`
+	CompletionTime uint64 `json:"endTime"`
+	Message        string `"json:message"`
```

Comment:
Please use `"json:errorMessage"` to be consistent with the NodeManagementPolicyStatus.